loopy/statistics.py

```diff
@@ -1316,14 +1317,24 @@ def map_subscript(self, expr):
         except AttributeError:
             var_tags = frozenset()
 
+        is_global_temp = False
         if name in self.knl.arg_dict:
             array = self.knl.arg_dict[name]
```

Comment:
Isn't this code path missing a `self.rec`?

Reply:
If `name` is in `arg_dict`, we still may need to count the global access before recursing; that happens below. This could be made less compact with more conditionals, which might be slightly easier to comprehend, but for efficiency I avoided duplicated logic/assignments and unnecessary checks. I'll describe what it is supposed to be doing, and if you still think there's a problem or have a clearer way to organize the various conditions, let me know.

Each of the three branches of this conditional are supposed to either (1) recurse on `expr.index`, *or* (2) assign something to `array` and set `is_global_temp` if necessary (recursing happens later):

- In the first case, we have found the subscript name in `arg_dict`, so it assigns `array` and leaves `is_global_temp=False`.

- In the second case, we have found the subscript name in `temporary_variables`. It recurses if the address space is *not* global, and otherwise assigns to `array` and sets `is_global_temp=True`.

- In the `else` case, we haven't found the subscript in `array_args` or `temporary_variables`, so it recurses.

After this three-part conditional, if we didn't already recurse, the next if-statement only applies if the subscript is *not* a global temporary; if this is the case, it checks to make sure `array` is a `lp.ArrayArg` and, if it is not, we recurse.

After that, if we haven't yet recursed, we're dealing with either a global temporary, or a `lp.ArrayArg` that was found in `arg_dict`, so we proceed as before and count it.

Reply:
Thanks for explaining. The control flow here is a bit more convoluted than I would like, but I agree that it's right.